spec/spec.md

```diff
@@ -209,6 +209,27 @@ The type of the `If` expression is the common type between `A` and `B`.
 **Expression kind**: r-value
 **Uses**: `cond`, `body`, and - if present - `else`
 
+#### `While`
+
+```grammar
+expr += (While cond:<expr> body:<expr>)
+```
+
+Repeatedly evaluates `body`, as long as `cond` evaluates to `true`. Both
+`body` and `cond` are part of a new scope.
+
+Let `C` be the type of `cond` and let `T` be the type of `body`. An error is
+reported if:
+* `C` is not `bool`
+* `T` is not `unit`
+
+The type of a `While` expression depends on the `cond` expression. If `cond`
+is the built-in `true` literal `(Ident "true")`, then the expression is of
+type `void`, otherwise it is of type `unit`.
```

Comment:
should this document the expected behaviour, that if the condition evaluates to true after constant expression evaluation then the loop is `void`, with a note saying the implementation doesn't yet do that?

Reply:
A note mentioning the direction is a good idea, though I would leave the mentioning implementation out of it, so that the specification stays independent from the implementation.

Unrelated to the note, but macros and constant expressions are going to cause us some trouble, semantics wise. If compile-time evaluation is only partially defined (like it is in NimSkull), then the static semantics of the language become implementation defined, which I consider quite problematic. Having *all* dynamic behaviour be well-defined is going to be a challenge, I believe.